### PR TITLE
BUG: Fix leap second table download triggered on import

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to address the problem of leap-second table being downloaded on module import, causing caching problems in cluster computing sharing the same cache.

Disclaimer: I did not benchmark the performance hit. Is performance even an issue we need to consider for this patch?

Address the problem reported in https://github.com/astropy/astropy/issues/9479#issuecomment-610663029

### CODE REVIEW

**Comment by pllim** [comment]
Only used in that module.

File: CHANGES.rst [hunk 1]
```diff
@@ -343,6 +343,8 @@ astropy.coordinates
 - Ensure that for size-1 array ``SkyCoord`` and coordinate frames
   the attributes also properly become scalars when indexed with 0.
   [#10113]
+- Delay download of leap-second table in ``earth_orientation.py`` until
```
Comment (mhvk) [inline CHANGES.rst:346]: Needs empty line, I think...

**Comment by mhvk** [comment]
If this works, I think it may be possible to do the same in `time.formats`, so that even importing the `time` module wouldn't download the `LeapSecond` table immediately...

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/issues/9479#issuecomment-611019144 and the comments above -- Would be nice if @weaverba137 can take this PR for a spin and see where it crashes next.

I'd rather we fix a real use case than go hunting for all `Time()` calls.

File: CHANGES.rst (hunk 1 above)
Comment (pllim) [reply CHANGES.rst:346]: Ops, good catch! I'll add it if we decide to proceed with this PR.

**Comment by weaverba137** [comment]
I will try to run a test when I can.  I'm pretty slammed today.

**Comment by pllim** [comment]
These might be in scope too:
```
astropy/coordinates/builtin_frames/lsr.py:J2000 = Time('J2000')
astropy/coordinates/builtin_frames/utils.py:EQUINOX_J2000 = Time('J2000', scale='tt')
astropy/coordinates/builtin_frames/utils.py:EQUINOX_B1950 = Time('B1950', scale='tt')
astropy/coordinates/builtin_frames/utils.py:DEFAULT_OBSTIME = Time('J2000', scale='tt'
```

**Comment by mhvk** [comment]
Of course, those are a lot trickier, since an actual time instance is wanted. I'm starting to think the leap second loading itself should be moved away from `__new__` to where it might actually get used (I don't quite recall why I didn't start with that...)

**Comment by pllim** [comment]
The one in LSR has a vague "for speed" comment but it is not used in the module. Maybe @adrn can explain why it was added in #6219 and whether it is still relevant.

https://github.com/astropy/astropy/blob/b505c4b5d9b673adf5b756208ed30cd1f02608dd/astropy/coordinates/builtin_frames/lsr.py#L18-L19

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/10125#issuecomment-611178990

Ah, @mhvk , if you can nip it in the bud, then maybe this PR won't be needed. I'll put this on hold.

**Comment by adrn** [comment]
> The one in LSR has a vague "for speed" comment but it is not used in the module. Maybe @adrn can explain why it was added in #6219 and whether it is still relevant.

I don't remember. Try removing and see if tests still pass? 🤷‍♂

**Comment by mhvk** [comment]
@pllim - not sure that I'll succeed - I don't remember much about my trials, but do remember it was an enormous hassle to avoid circular dependencies...

**Comment by bsipocz** [comment]
> Ah, @mhvk , if you can nip it in the bud, then maybe this PR won't be needed. I'll put this on hold.

What about going gradually, and get this in now and instead of holding up the PR, just revert if/when the other approach works out and gets in.

**Comment by mhvk** [comment]
I'm fine with replacing the jds here with actual values (no need for the lazy loading), but think it is unlikely it actually solves the problem, so the changelog entry may need to be adjusted.

**Comment by pllim** [comment]
> replacing the jds here with actual values 

That works for me. Will the values ever change due to leap seconds? Maybe we can even include it in `constants` and be done with it. I see a few flavors of `Time(epoch)` calls across the package.

**Comment by pllim** [comment]
It is too bad we cannot use [PEP 562](https://www.python.org/dev/peps/pep-0562/) yet. The ones in `builtin_frames/utils.py` are tricky.

**Comment by pllim** [comment]
This is looking pretty hacky... 😬

**Comment by mhvk** [comment]
Great, now at least it is clear where the problems lie.  Will try to find some time to see if I can instead move the download further down.

**Comment by taldcroft** [comment]
Probably the coffee still isn't working, but I'm confused about this approach and I don't see how this is preventing making new Time objects at import time, which if I understand can trigger a download.  I made a file `lazy.py`:
```
def func(arg):
    print(f'func called: {arg}!')
    return arg

class TimeAttribute:
    def __init__(self, arg):
        print(f'In TimeAttribute: {arg}')

class Blah:
    attr = TimeAttribute(arg=func('first'))

    def meth(arg=func('second')):
        pass
```
Then
```
>>> import lazy
func called: first!
In TimeAttribute: first
func called: second!
```

**Comment by pllim** [comment]
@taldcroft , I didn't know Python actually runs the nested code on import. 🤔  I'll have to double check.

**Comment by pllim** [comment]
I think the trick is to make them class properties. If I changed @taldcroft 's `lazy.py` to the following:

```python
class Blah:
    @property
    def attr(self):
        return TimeAttribute(arg=func('first'))

    def meth(arg=func('second')):
        pass
```

I only get:
```python
>>> import lazy
func called: second!
```

**Comment by taldcroft** [comment]
I do hope that @mhvk can find a lower-level solution.  It does not seem like there should be any requirement to access the latest leap seconds just to *create* a Time object.  It's only for converting.

A lower-level solution also helps other projects that may be creating Time objects on import.  It seems that importing any part of astropy should never trigger or rely on network access.

**Comment by pllim** [comment]
Well, I am stuck anyway. I went down the rabbit hole of frame attributes. Changing them to class property broke everything. There is some magic going on with class level attributes and there is no easy way to lazy load them. 😢

**Comment by pllim** [comment]
I am closing this PR. I think this is a dead end. Sorry for the noise!

**Comment by mhvk** [comment]
@pllim - thanks for the valiant attempt!  I agree that there's a need for a lower-level solution, and it should be do-able (though it is not like I didn't try initially; am hoping that a fresh look will help...)

**Pull request closed without merging**